### PR TITLE
MWPW-156612 fragment merch-card variant layouts

### PR DESCRIPTION
**Pull request by npeltier**
- move from merch-card.js code each layout code related to a given variant under variants/`variant` class
- move from global.css.js css related to a given variant under variants/`variant`.css.js
- move from merch-card.css.js wc css related to a given variant under variants/`variant`.variantStyle
- uses variants/variants as an index to get related class from related variant, and wc style too,
- uses variants/VariantLayout, superclass of all variants, to insert css if card is used, and hold common code

Resolves: https://jira.corp.adobe.com/browse/MWPW-156612

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/?martech=off
- After: https://MWPW-156612--milo--npeltier.hlx.page/?martech=off

- Before: https://main--milo--adobecom.hlx.live/docs/library/kitchen-sink/merch-card?martech=off&georouting=off
- After: https://MWPW-156612--milo--npeltier.hlx.live/docs/library/kitchen-sink/merch-card?martech=off&georouting=off

tested kitchen sink with consolidated style in one tag done before first card is even parsed https://test-consolidate--milo--npeltier.hlx.live/docs/library/kitchen-sink/merch-card?martech=off&georouting=off but does not look like it greatly improves TBT

### CODE REVIEW

**Comment by npeltier** [comment]
will probably need to wait for https://github.com/adobecom/milo/pull/2798 but reviews can start

**Comment by npeltier** [comment]
(i also need to check why TBT is increased)

File: libs/features/mas/web-components/src/variants/mini-compare-chart.js
Comment (narcis-radu) [inline libs/features/mas/web-components/src/variants/mini-compare-chart.js:99]: nit: might be safer to use `...footerRows?.children` to avoid situations where the `footer-rows` element is not available.

File: libs/features/mas/web-components/src/variants/twp.css.js
Comment (narcis-radu) [inline libs/features/mas/web-components/src/variants/twp.css.js:9]: is RTL covered?

File: libs/features/mas/web-components/src/variants/twp.css.js
Comment (narcis-radu) [inline libs/features/mas/web-components/src/variants/twp.css.js:38]: nit: missing `;`

File: libs/features/mas/web-components/src/variants/twp.css.js
Comment (narcis-radu) [inline libs/features/mas/web-components/src/variants/twp.css.js:56]: I am looking at the order and I think declarations should follow mobile to large desktop or vice versa, however here we have desktop, large desktop, mobile, tablet, which might get confusing at some point when different CSS is added

File: libs/features/mas/web-components/src/variants/variant-layout.js
Comment (narcis-radu) [inline libs/features/mas/web-components/src/variants/variant-layout.js:10]: this has potential to break; let's say `VariantLayout.styleMap` is undefined, then your if clause will evaluate to true and it will execute `VariantLayout.styleMap[this.card.variant] = true;` which will throw an error.

File: libs/features/mas/web-components/src/variants/variant-layout.js
Comment (narcis-radu) [inline libs/features/mas/web-components/src/variants/variant-layout.js:83]: are these methods still needed then?

File: libs/deps/mas/merch-card.js
Comment (npeltier) [reply libs/deps/mas/merch-card.js:1633]: i'll merge @yesil 's PR that does that, but yes

File: libs/features/mas/web-components/src/variants/twp.css.js
Comment (npeltier) [reply libs/features/mas/web-components/src/variants/twp.css.js:9]: not yet, no this one is still under dev

**Review by 3ch023 — APPROVED** [review]

File: libs/features/mas/web-components/src/merch-card.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/merch-card.js:300]: This method is only applicable to `catalog` card. I wonder if it should be moved to `catalog.js`?

File: libs/features/mas/web-components/src/variants/catalog.css.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/catalog.css.js:30]: nit: add missing line

File: libs/features/mas/web-components/src/variants/image.css.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/image.css.js:26]: nit: add missing line

File: libs/features/mas/web-components/src/variants/plans.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/plans.js:48]:
```suggestion
    :host([variant='plans']) ::slotted([slot='heading-xs']) {
```

File: libs/features/mas/web-components/src/variants/plans.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/plans.js:47]: nit: add missing line

File: libs/features/mas/web-components/src/variants/product.css.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/product.css.js:28]: nit: add missing line

File: libs/features/mas/web-components/src/variants/segment.css.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/segment.css.js:25]: nit: add missing line

File: libs/features/mas/web-components/src/variants/special-offer.css.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/special-offer.css.js:28]: nit: add missing line

File: libs/features/mas/web-components/src/variants/special-offer.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/special-offer.js:44]: nit: add missing line

File: libs/features/mas/web-components/src/variants/twp.css.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/twp.css.js:7]: nit: add missing line

File: libs/features/mas/web-components/src/variants/twp.css.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/variants/twp.css.js:99]: nit: add indent

**Review by Axelcureno — CHANGES_REQUESTED** [review]
Some nits that could be ignored, but this one does need attention:
https://github.com/adobecom/milo/pull/2810/files#r1750613115

File: libs/features/mas/web-components/src/merch-card.js
Comment (npeltier) [reply libs/features/mas/web-components/src/merch-card.js:300]: had some issues there but you are right, i'll try again :)

**Review by Axelcureno — APPROVED** [review]

**Comment by afmicka** [comment]
@npeltier the style of merch-card with background opacity has changed. Was that intentional? 
your branch:
Screenshot 2024-09-11 at 14 37 22

actual: 
Screenshot 2024-09-11 at 14 37 34

**Comment by afmicka** [comment]
@npeltier Seeing also differences in width on mobile for special offers cards:
your branch: 
Screenshot 2024-09-11 at 14 54 43

act
Screenshot 2024-09-11 at 14 37 34
ual:

**Comment by npeltier** [comment]
@afmicka background image, special offers, and bacom page should be fixed now. The width issue on kitchen sink is due to style ordering in case of multiple type of cards inserted in the same section. i'm not sure what we want here & started to discuss this internally.

File: libs/features/mas/web-components/src/variants/variant-layout.js
Comment (yesil) [inline libs/features/mas/web-components/src/variants/variant-layout.js:4]: nit: I tend to use module scoped variants instead of static members.
Especially if they are meant to be `private`

File: libs/features/mas/web-components/src/variants/variant-layout.js
Comment (yesil) [inline libs/features/mas/web-components/src/variants/variant-layout.js:9]: nit: with above suggestion, it could be written as `if (!styleMap[this.card.variant]) {`

Comment (yesil) [reply libs/features/mas/web-components/src/variants/variant-layout.js:9]: since we append the style below, I would call the method `appendVariantStyles`

File: libs/features/mas/web-components/src/variants/variant-layout.js
Comment (yesil) [inline libs/features/mas/web-components/src/variants/variant-layout.js:19]: is `setTimeout` really needed?

File: libs/features/mas/web-components/src/variants/variant-layout.js
Comment (yesil) [inline libs/features/mas/web-components/src/variants/variant-layout.js:22]: Why this and the getters below do not stay in `merch-card.js` ? To me `VariantLayout` class looks not necessary.

File: libs/features/mas/web-components/test/merch-card.catalog.test.html.js
Comment (yesil) [inline libs/features/mas/web-components/test/merch-card.catalog.test.html.js:26]: if you feel,  remove `appendMiloStyles`, and maybe reference the styles in question in the test page.

File: libs/features/mas/web-components/src/variants/catalog.js
Comment (yesil) [inline libs/features/mas/web-components/src/variants/catalog.js:9]: do we need this constructor?

File: libs/features/mas/web-components/src/variants/ccd-action.js
Comment (yesil) [inline libs/features/mas/web-components/src/variants/ccd-action.js:10]: suggestion: `getGlobalStyles`?

File: libs/features/mas/web-components/src/variants/catalog.js
Comment (yesil) [inline libs/features/mas/web-components/src/variants/catalog.js:79]: I think we should reference to styles always in plural and since styles is already not used, why not simply use: `static styles = css`

**Review by yesil — COMMENTED** [review]
Looks like more styles can be moved from global.css.js to the merch-card.css.js  and some suggestion on the naming.
Overall it looks good to me, will wait for the comments before approval.

File: libs/features/mas/web-components/src/global.css.js
Comment (npeltier) [reply libs/features/mas/web-components/src/global.css.js:1]: sure, i thought we would wait for the move to "optimize" but i'll do that :)

File: libs/features/mas/web-components/src/variants/variant-layout.js
Comment (npeltier) [reply libs/features/mas/web-components/src/variants/variant-layout.js:19]: i  _think_ i did save some TBT with this yes

File: libs/features/mas/web-components/src/variants/variant-layout.js
Comment (npeltier) [reply libs/features/mas/web-components/src/variants/variant-layout.js:22]: Because such methods affect layout. And layout is now deferred to the VariantLayout classes. Sometimes some are being overriden.

File: libs/features/mas/web-components/src/variants/catalog.js
Comment (npeltier) [reply libs/features/mas/web-components/src/variants/catalog.js:9]: yes we do! the constructor adds the style if not there

Comment (yesil) [reply libs/features/mas/web-components/src/variants/catalog.js:9]: but when you don't declare, it should call the one from VariantLayout no?

**Comment by afmicka** [comment]
issues seem fixed. Will mark it verified. 
 @Roycethan could you please also have a look just in case? If all good on your end, please add Ready for stage.

**Review by yesil — APPROVED** [review]

File: libs/features/mas/web-components/test/merch-card.catalog.test.html.js
Comment (npeltier) [reply libs/features/mas/web-components/test/merch-card.catalog.test.html.js:26]: finally someone will do that later, as it's more work than i was expected :)

File: libs/features/mas/web-components/src/global.css.js
Comment (npeltier) [reply libs/features/mas/web-components/src/global.css.js:1]: finally, same as for web tests, i'll postpone it as it's not that easy neither :)

File: libs/features/mas/web-components/src/variants/catalog.js
Comment (npeltier) [reply libs/features/mas/web-components/src/variants/catalog.js:9]: i wanted to shout NO, but i tried and yes it works :) sorry to be such a java guy :)

**Comment by npeltier** [comment]
@afmicka postponed heavy styling changes to later PR (cc @yesil ) so no additional testing on your hand is needed :)
it's all in the hands of @Roycethan now :)

**Pull request merged**
